Запрос рекомендации 4/4

DenisIgnatenko opened the pull request:
В четвертой части включена третья.

DenisIgnatenko committed:
feat: 1/4: Запрос рекомендации / Запросить рекомендацию

DenisIgnatenko committed:
feat: 2/4: Запрос рекомендации / Получить запросы рекомендаций с фильтрами. Можно фильтровать по статусу (PENDING, ACCEPTED, REJECTED), по ID запрашивающего и запрашиваемого (например http://localhost:8080/api/recommendations?requesterId=1), и по дате startDate, endDate (например ttp://localhost:8080/api/recommendations?startDate=). Так же можно комбинировать запросы (http://localhost:8080/api/recommendations?status=PENDING&requesterId=2)

DenisIgnatenko committed:
feat: 4/4: Запрос рекомендации / Отклонить запрос на рекомендацию. Можно отклонить запрос на рекомендацию, отправив в на /api/recommendations/{id}/reject, например http://localhost:8080/api/recommendations/11/reject/.

Если запрос имеет статус REJECTED или ACCEPTED - его нельзя будет отклонить. Если PENDING - то вернется json с обновленной записью, включая причину отказа. Если причина отказа в BODY запроса не была передана - вернется ошибка с сообщением о необходимости указать причину отказа.

drewbal commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,134 @@
+    private final RecommendationRequestMapper recommendationRequestMapper;
+    private final UserRepository userRepository;
+    private final SkillRepository skillRepository;
+    private final SkillRequestRepository skillRequestRepository;
```
Сервис напрямую работает только со своим репозиторием. С другими репозиториями общается через их сервисы.

drewbal commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,134 @@
+            if (existingSkillsCount != skillsIds.size()) {
+                throw new IllegalArgumentException("Некоторых скиллов нет в базе данных");
+            }
+        }
```
Валидацию можно вынести в отдельный класс, чтобы не прописывать такие конструкции большие конструкции в каждом методе и облегчить переиспользование

drewbal commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,134 @@
+                skillRequestRepository.save(skillRequest);
+
+                savedRequest.getSkills().add(skillRequest);
+            }
```
Логика выше (про skillIds) больше принадлежит сервису, связанному с сущностью SkillRequest, а не RecommendationRequest, так что можно перенести туда, гранулировать функционал, повысить масштабируемость и облегчить тестируемость :)

RuslanZinovyev replied:
Что это за логика и как она логически связана с данным методом create()?

DenisIgnatenko replied:
В условии задачи сказано, что нужно отдельно сохранить все запрашиваемые скиллы в таблицу skill_request (задача 8).

Я создал отдельный сервис SkillRequestService и вынес туда логику добавления каждого из запрашиваемых для рекомендации скиллов в таблицу skill_request.

Еще более упростил код, создал метод, который берет список скиллов и добавляет их в базу.

drewbal commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,134 @@
+    private final SkillRepository skillRepository;
+    private final SkillRequestRepository skillRequestRepository;
+
+    public RecommendationRequestDto create(RecommendationRequestDto dto) {
```
Еще можно повесить @Transactional, т.к. в рамках метода мы делаем несколько операций с БД и нам надо их сделать атомарно, одной транзакцией. Это уберет риск частично исполненного запроса, если с одной из бд у нас возникнут проблемы, например

drewbal commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,134 @@
+                        .isAfter(filter.getEndDate()))
+                .map(recommendationRequestMapper::toDto)
+                .collect(Collectors.toList());
+    }
```
Фильтрацию лучше делать через отдельный интерфейс, классы фильтров и их список. Таким образом, при добавлении нового фильтра мы его добавим в дто всех фильтров, создадим класс, но не будем ничего менять в текущем методе. 
sOlid :)

Работу с фильтрами Влад объяснял в видео по MapStruct, а также есть подробный вопрос в workshop видео следующего спринта.

drewbal commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,134 @@
+        RecommendationRequest request = recommendationRequestRepository.findById(id)
+                .orElseThrow(() -> new GlobalExceptionHandler.RecommendationRequestNotFoundException(
+                        "Запрос на рекомендацию с таким id не найден"));
+        return recommendationRequestMapper.toDto(request);
```
Валидация

drewbal commented on src/test/java/school/faang/user_service/service/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,466 @@
+        requester.setId(1L);
+
+        User receiver = new User();
+        receiver.setId(2L);
```
У тебя сетап для каждого теста +/- одинаковый создается. Пропиши его один раз в @BeforeEach void setUp и избавься от рутины. А в тестах можно разово через сеттеры менять требуемые контекстом параметры.

И еще для создания объектов могу посоветовать попробовать @Builder из Lombok. Мне очень понравилось. Можно даже налету создавать объект хоть с единственным необходимым для теста параметром

drewbal reviewed (commented):
Денис, классный код. Почернул от тебя "advanced" фишки, которых не было в лекциях еще :))

У меня несколько предложений на улучшение логики, масштабируемости, SOlid-ола :)

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,53 @@
+        return ResponseEntity.ok(requests);
+    }
+
+    @GetMapping("{id}")
```
Это не проблема ибо Spring может это решить сам, но если ты в одном ендпоинте добаил слэш то делай это и для других /{id}

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,53 @@
+
+    @GetMapping
+    public ResponseEntity<List<RecommendationRequestDto>> getRecommendationRequests(
+            @Valid @ModelAttribute RequestFilterDto filter) {
```
ModelAttribute не очень очевидный выбор для REST приложения, в MVC приложении в нем больше смысла.

RuslanZinovyev commented:
Разреши все комментарии выше включая рекомендации что скинул Андрей и мы закроем данный PR!

DenisIgnatenko committed:
fix: работа над комментариями по коду.
1. Убрал взаимодействие RecommendationRequestService с другими репозиториями напрямую. Создал сервисы SkillRequestService, UserService. Настроил общение через них
2. Вынес валидацию в отдельный класс. Теперь вся валидация выполняется там
3. Создал сервис SkillRequestService для работы со скиллами.
4. Повесил @transactional на методы, которые делают несколько операций с БД
5. Перелопатил всю фильтрацию. Теперь каждая выполняется в своем классе, который наследуется от интерфейса RequestFilter
6. Переработал тесты. Добавил метод setUp с аннотацией @beforeEach чтобы избежать дублирования кода
7. Поправил контроллер с некорректным неймингом эндпоинтов.

YTXIRE commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/recommendations")
```
Добавь версионирование в контроллер.
А именно, укажи в пути к контроллеру /api/v1/recommendations

YTXIRE commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,53 @@
+public class RecommendationRequestController {
+    private final RecommendationRequestService recommendationRequestService;
+
+    @PostMapping("/request")
```
Не совсем понятное название метода. Предлагаю заменить на что - то более понятное. Либо вообще убрать  /request, поскольку и так у тебя будет работать логика, но уже более понятно за что метод будет отвечать.

YTXIRE commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,53 @@
+    private final RecommendationRequestService recommendationRequestService;
+
+    @PostMapping("/request")
+    public RecommendationRequestDto requestRecommendation(
```
Оберни ответ в ResponseEntity

YTXIRE commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,53 @@
+        return ResponseEntity.ok(dto);
+    }
+
+    @PatchMapping("/{id}/reject")
```
Я думаю, что проще использовать PutMapping. Patch уже редко используется. И будет проще использовать put для обновления.

YTXIRE commented on src/main/java/school/faang/user_service/dto/RecommendationRequestDto.java:
```diff
@@ -0,0 +1,26 @@
+public class RecommendationRequestDto {
+    private Long id;
+
+    @NotEmpty(message = "Сообщение не должно быть пустым")
```
Если добавляешь такую валидацию на поле dto, то стоит и на остальные поля добавить.
А вместо NotEmpty, которая проверяет, что поле просто не пустое, можно использовать @NotBlank, которая првоерит, что там не стоит пробела. По хорошему еще и добавить валидацию длинны.

YTXIRE commented on src/main/java/school/faang/user_service/dto/RequestFilterDto.java:
```diff
@@ -0,0 +1,25 @@
+    private LocalDateTime startDate;
+    private LocalDateTime endDate;
+
+    @Min(value = 1, message = "Ошибка в ID. Номер должен быть больше 0")
```
Можно вместо Min использовать аннотацию @Positive.
Ошибки стайся писать на английском.

YTXIRE commented on src/main/java/school/faang/user_service/dto/RequestFilterDto.java:
```diff
@@ -0,0 +1,25 @@
+    @Min(value = 1, message = "Ошибка в ID. Номер должен быть больше 0")
+    private Long requesterId;
+
+    @Min(value = 1, message = "Ошибка в ID. Номер должен быть больше 0")
```
Можно вместо Min использовать аннотацию @positive.
Ошибки стайся писать на английском.

YTXIRE commented on src/main/java/school/faang/user_service/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,32 @@
+        return ResponseEntity.badRequest().body(exception.getMessage());
+    }
+
+    @ExceptionHandler(Exception.class)
```
Лучше указывать конкретное исключение, нежели целую группу.

DenisIgnatenko replied:
Да, я просто пока не догадался какой именно. Я описал те, что понимаю и сделал общий на другие, пока не разберусь на что заменить

YTXIRE commented on src/main/java/school/faang/user_service/filter/EndDateFilter.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.stream.Stream;
+
+@Component
+public class EndDateFilter implements RequestFilter {
```
Используй интефейс фильтра из папки фильтров, который в мастере.

YTXIRE commented on src/main/java/school/faang/user_service/filter/ReceiverIdFilter.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.stream.Stream;
+
+@Component
+public class ReceiverIdFilter implements RequestFilter {
```
Используй интефейс фильтра из папки фильтров, который в мастере.

YTXIRE commented on src/main/java/school/faang/user_service/filter/RecommendationRequestFilterManager.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.stream.Stream;
+
+@Component
+public class RecommendationRequestFilterManager {
```
А есть ли смысл в этом классе?

YTXIRE commented on src/main/java/school/faang/user_service/filter/RequestFilter.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.util.stream.Stream;
+
+public interface RequestFilter {
```
Для этого есть уже готовый фильтр - интерфейс с дженериками.

YTXIRE commented on src/main/java/school/faang/user_service/filter/RequesterIdFilter.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.stream.Stream;
+
+@Component
+public class RequesterIdFilter implements RequestFilter {
```
Используй интефейс фильтра из папки фильтров, который в мастере.

YTXIRE commented on src/main/java/school/faang/user_service/filter/SkillTitleFilter.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.stream.Stream;
+
+@Component
+public class SkillTitleFilter implements RequestFilter {
```
Используй интефейс фильтра из папки фильтров, который в мастере.

YTXIRE commented on src/main/java/school/faang/user_service/filter/StartDateFilter.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.stream.Stream;
+
+@Component
+public class StartDateFilter implements RequestFilter {
```
Используй интефейс фильтра из папки фильтров, который в мастере.

YTXIRE commented on src/main/java/school/faang/user_service/filter/StatusFilter.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.stream.Stream;
+
+@Component
+public class StatusFilter implements RequestFilter {
```
Используй интефейс фильтра из папки фильтров, который в мастере.

YTXIRE commented on src/main/java/school/faang/user_service/mapper/RecommendationRequestMapper.java:
```diff
@@ -0,0 +1,36 @@
+        }
+        return skillRequests.stream()
+                .map(skillRequest -> skillRequest.getSkill().getId())
+                .collect(Collectors.toList());
```
Можно сразу указать .toList();

YTXIRE commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,103 @@
+        recommendationRequestValidator.validateRequestFrequency(dto.getRequesterId(), dto.getReceiverId());
+
+        List<Long> skillsIds = dto.getSkills();
+        if (skillsIds != null && !skillsIds.isEmpty()) {
```
Эту проверку можно спрятавь в общий метод валидации или сделать валидацию через dto.

YTXIRE commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,103 @@
+    public List<RecommendationRequestDto> getRequests(RequestFilterDto filter) {
+        List<RecommendationRequest> requests = recommendationRequestRepository.findAll();
+
+        Stream<RecommendationRequest> stream = requests.stream();
```
Данную операцию можно объединить с операцией выше, а именно получение данных из базы.

YTXIRE commented on src/main/java/school/faang/user_service/service/UserService.java:
```diff
@@ -0,0 +1,17 @@
+
+    public User getUserById(Long userId) {
+        return userRepository.findById(userId)
+                .orElseThrow(() -> new IllegalArgumentException("Пользователя с id " + userId + " не существует"));
```
Для таких моментов создается исключение Empty и кладется в пакет exceptions.

YTXIRE commented on src/main/java/school/faang/user_service/service/SkillRequestService.java:
```diff
@@ -0,0 +1,34 @@
+
+    public Skill getSkillById(Long skillId) {
+        return skillRepository.findById(skillId)
+                .orElseThrow(() -> new IllegalArgumentException("В базе данных нет скилла с id: " + skillId));
```
Для таких моментов создается исключение Empty и кладется в пакет exceptions.

YTXIRE commented on src/test/java/school/faang/user_service/service/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,537 @@
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+class RecommendationRequestServiceTest {
```
Почему тесты только на сервис?
У тебя же есть еще валидатор, контроллер. И сервисов у тебя не один. У тебя еще есть разные фильтры.

RuslanZinovyev commented:
Нужно разрешить все конфликты.

DenisIgnatenko committed:
fix: проработал все замечания, кроме использования интерфейса из мастера. После коммита подтяну и доделаю.

DenisIgnatenko committed:
Merge branch 'Dragon-master-stream7' into Dragon-master-stream7-BJS2-40456

# Conflicts:
#	src/main/java/school/faang/user_service/dto/RejectionDto.java
#	src/main/java/school/faang/user_service/dto/RequestFilterDto.java
#	src/main/java/school/faang/user_service/exception/GlobalExceptionHandler.java
#	src/main/java/school/faang/user_service/service/UserService.java

DenisIgnatenko committed:
fix: переработал фильтры под общий интерфейс с дженериками

DenisIgnatenko committed:
merge branch Dragon-master-stream7 into Dragon-master-stream7-BJS2-40456

DenisIgnatenko committed:
Updated my branch with last master brand updates

DenisIgnatenko committed:
feat: Добавлены тесты:

1. SkillRequestServiceTest
2. RecommendationRequestControllerTest
3. RecommendationRequestValidatorTest
4. RecommendationRequestMapperTest

DenisIgnatenko committed:
Merge branch 'refs/heads/Dragon-master-stream7' into Dragon-master-stream7-BJS2-40456

# Conflicts:
#	src/main/java/school/faang/user_service/service/UserService.java

YTXIRE commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/v1/recommendations")
```
Подлей мастер и убери из запроса /api/v1
Теперь это не нужно указывать в контроллере, это вынесено в конфиг

YTXIRE commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,53 @@
+
+    @PutMapping("/{id}/reject")
+    public ResponseEntity<RecommendationRequestDto> rejectRequest(
+            @PathVariable Long id, @Valid @RequestBody RejectionDto rejection) {
```
Повесил бы сразу валидацию на PathVariable

YTXIRE commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @GetMapping("/{id}")
+    public ResponseEntity<RecommendationRequestDto> getRecommendationRequest(@PathVariable Long id) {
```
Повесил бы сразу валидацию на PathVariable

YTXIRE commented on src/main/java/school/faang/user_service/exception/GlobalExceptionHandler.java:
```diff
@@ -2,12 +2,32 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
 
-@ControllerAdvice
+@RestControllerAdvice
```
Смысл менять?

DenisIgnatenko replied:
А я подумал, что будет удобно, если не нужно будет добавлять аннотацию @ResponseBody каждому методу, поскольку она автоматически включена в @RestControllerAdvice, но раз я все равно использую ResponseEntity (типа явно сериализирую), то оно и с @ControllerAdvice нормально работает.

DenisIgnatenko committed:
feat:
1. RecommendationRequestController: добавил валидацию  на pathVariable.
2. Смерджил новый мастер запроса /api/v1
3. Скоректировал тесты

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,57 @@
+@RestController
+@RequestMapping("/recommendationrequest")
+@RequiredArgsConstructor
+@Validated
```
Для чего нужна эта аннотация? Ты где то используешь групповую валидацию?

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/recommendationrequest")
```
`/recommendation-request`

RuslanZinovyev commented on src/main/java/school/faang/user_service/exception/GlobalExceptionHandler.java:
```diff
@@ -1,13 +1,37 @@
+import java.util.stream.Collectors;
+
 @ControllerAdvice
 public class GlobalExceptionHandler {
```
Не вижу логгера и логирования стэк трейса, как разработчики поймут что конкретно пошло не так и как будут расследовать подобные проблемы?

DenisIgnatenko replied:
А я тут его еще не использовал, думал подтянуть логирование на следующих задачах. Но сейчас разберусь как добавить.

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,103 @@
+        User requester = userService.getUserById(dto.getRequesterId())
+                .orElseThrow(() -> new EntityNotFoundException("Requester with ID " + dto.getRequesterId() + " not found"));
+        User receiver = userService.getUserById(dto.getReceiverId())
+                .orElseThrow(() -> new EntityNotFoundException("Requester with ID " + dto.getRequesterId() + " not found"));
```
Должно быть

```
User receiver = userService.getUserById(dto.getReceiverId())
        .orElseThrow(() -> new EntityNotFoundException("Receiver with ID " + dto.getReceiverId() + " not found"));
```

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,103 @@
+
+        RecommendationRequest recommendationRequest = recommendationRequestMapper.toEntity(dto);
+
+        if (recommendationRequest.getSkills() == null) {
```
А если getSkills() != null ты просто ничего не добавляешь? 

`recommendationRequest.setSkills(Optional.ofNullable(recommendationRequest.getSkills()).orElse(new ArrayList<>()));`

RuslanZinovyev commented on src/main/java/school/faang/user_service/dto/RecommendationRequestDto.java:
```diff
@@ -0,0 +1,43 @@
+    private RequestStatus status;
+
+    @NotEmpty(message = "Skill list must not be empty")
+    private List<Long> skills;
```
`skillIdentifiers` более подходящее имя для данного поля

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/SkillRequestService.java:
```diff
@@ -0,0 +1,35 @@
+    private final SkillRepository skillRepository;
+
+    public SkillRequest createSkillRequest(Skill skill, RecommendationRequest request) {
+        SkillRequest skillRequest = new SkillRequest();
```
SkillRequest имплементирует паттер Builder, используй его преимущества при создании объекта.

RuslanZinovyev commented on src/main/java/school/faang/user_service/validator/RecommendationRequestValidator.java:
```diff
@@ -0,0 +1,62 @@
+
+    public void validateUsersExistence(User requester, User receiver) {
+        if (requester == null) {
+            throw new IllegalArgumentException("Пользователя, запрашивающего рекомендацию не существует");
```
Мы не используем русский язык в написании кода практически никогда. Используй английский. Исправь везде, грамматику никто проверять не будет =)

RuslanZinovyev commented on src/test/java/school/faang/user_service/mapper/RecommendationRequestMapperTest.java:
```diff
@@ -0,0 +1,114 @@
+
+    @Test
+    void testToDto_Success() {
+        RecommendationRequest entity = new RecommendationRequest();
```
Используй паттерн Builder

DenisIgnatenko replied:
Блин, да, ты прав

RuslanZinovyev commented on src/test/java/school/faang/user_service/validator/RecommendationRequestValidatorTest.java:
```diff
@@ -0,0 +1,133 @@
+
+    @Test
+    void testValidateRequestFrequency_NoPreviousRequest() {
+        Long requesterId = 1L;
```
Создай константу если используешь переменную несколько раз.

DenisIgnatenko committed:
Merge branch 'refs/heads/Dragon-master-stream7' into Dragon-master-stream7-BJS2-40456

# Conflicts:
#	src/main/java/school/faang/user_service/service/UserService.java

DenisIgnatenko committed:
fix: Работа с комментариями
1. Поменял названия эндпоинтов на более аккуратные
2. Добавил логгирование в GlobalExceptionHandler
3. Поменял объявление экземпляров классов с использованием builder
4. Поменял описание ошибок с русского на английский
5. Немного поменял логику добавления скиллов, запрошенных на рекомендацию при создании рекомендации.
6. поправил мелкие ошибки в коде.
